### PR TITLE
Tensor str

### PR DESCRIPTION
**Pull request by liufengwei0103**
Tensor str

效果图：
int 类型：
截屏2021-08-09 下午7 22 30
int mode(val == np.ceil(val))，科学计数法：
截屏2021-08-09 下午7 24 03
int mode, 非科学计数法：
截屏2021-08-09 下午7 24 26
浮点类型：
截屏2021-08-09 下午7 24 45
浮点类型，科学计数法：
截屏2021-08-09 下午7 25 14
压缩数据形式：
截屏2021-08-09 下午7 27 21
consistent tensor打印（暂时不能打印数据）：
截屏2021-08-11 下午2 45 39
local tensor以及它的meta打印：
截屏2021-08-11 下午2 31 42
consistent tensor以及meta打印：
截屏2021-08-11 下午2 57 23

 @strint

### CODE REVIEW

File: python/oneflow/framework/tensor_str.py [hunk 1]
```diff
@@ -13,18 +13,215 @@
...
+PRINT_OPTS = __PrinterOptions()
+
+
+def _convert_to_local_tensor(self):
+    # consistent to local
+    if self.is_consistent:
+        placement = flow.placement("cpu", {0: [0]})
+        sbp = flow.sbp.broadcast
+        # TODO: delete `to("cuda")` after supporting cpu data broadcast
+        self = self.to("cuda").to_consistent(placement, sbp).to_local()
```
Comment (wyg1997) [inline python/oneflow/framework/tensor_str.py:43]: 目前不支持 cpu -> cpu 的broadcast，这里暂时要先to_cuda一下

File: python/oneflow/framework/tensor_str.py [hunk 2]
```diff
@@ -13,18 +13,215 @@
...
+        self.int_mode = True
+        self.sci_mode = False
+        self.max_width = 1
+        self.random_sample_num = 50
+
+        with flow.no_grad():
+            tensor_view = tensor.reshape([-1])
+
+        if not self.floating_dtype:
+            # TODO: verify int type tensor
```
Comment (wyg1997) [inline python/oneflow/framework/tensor_str.py:59]: 还不能构造出int类型的ConsistentTensor，这里的逻辑需要验证下

Comment (wyg1997) [reply python/oneflow/framework/tensor_str.py:59]: @clackhan @daquexian 现在能构造int类型的ConsistentTensor吗？我们上周试的时候，好像在Dispatch那里没有支持int类型的

File: python/oneflow/framework/tensor_str.py [hunk 3]
```diff
@@ -13,18 +13,215 @@
...
+
+def _tensor_str(self, indent):
+    summarize = self.numel() > PRINT_OPTS.threshold
+    if self.dtype is flow.float16:
+        self = self.float()
+
+    # TODO: convert to local before slicing consistent tensor
+    # self = _convert_to_local_tensor(self)
+    if self.is_consistent:
+        return "[]"
```
Comment (wyg1997) [inline python/oneflow/framework/tensor_str.py:213]: `[...]` 这样吧，要不看着像0shape Tensor

Comment (liufengwei0103) [reply python/oneflow/framework/tensor_str.py:213]: > `[...]` 这样吧，要不看着像0shape Tensor

有道理

File: python/oneflow/test/modules/test_tensor_str.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+import oneflow as flow
+import oneflow.unittest
+
+
+def _test_tensor_str(test_case, device):
+    # int dtype
+    tensor = flow.tensor([[1, 2, 3], [4, 5, 6]], device=flow.device(device))
+    tensor_str = str(tensor)
+    test_case.assertTrue('5' in tensor_str)
```
Comment (daquexian) [inline python/oneflow/test/modules/test_tensor_str.py:31]: 这里多 check 一些吧，1 2 3 4 5 6 都可以 check

File: python/oneflow/framework/tensor_str.py [hunk 5]
```diff
@@ -13,18 +13,204 @@
...
+    return "[" + tensor_str + "]"
+
+
+def _tensor_str(self, indent):
+    summarize = self.numel() > PRINT_OPTS.threshold
+    if self.dtype is flow.float16:
+        self = self.float()
+
+    # TODO: convert to local before slicing consistent tensor
+    # self = _convert_to_local_tensor(self)
```
Comment (daquexian) [inline python/oneflow/framework/tensor_str.py:200]: 这里不对，应该是先切片再 to local

Comment (liufengwei0103) [reply python/oneflow/framework/tensor_str.py:200]: > 这里不对，应该是先切片再 to local

这是之前的方案，之后会删掉。

File: python/oneflow/test/modules/test_tensor_str.py (hunk 4 above)
Comment (liufengwei0103) [reply python/oneflow/test/modules/test_tensor_str.py:31]: > 这里多 check 一些吧，1 2 3 4 5 6 都可以 check

好的

File: oneflow/core/functional/tensor_index.cpp [hunk 6]
```diff
@@ -66,6 +66,7 @@ Maybe<void> PrepareSliceIndices(const TensorIndex& index, const Shape& shape,
     CHECK_LT_OR_RETURN(dim, ndims) << "Invalid index for tensor of dimension " << ndims;
     if (index_item.IsSlice()) {
       const auto& slice = index_item.slice();
+      CHECK_GT_OR_RETURN(slice.step(), 0) << "Step must be greater than zero.";
```
Comment (liufengwei0103) [inline oneflow/core/functional/tensor_index.cpp:69]: 这个位置相比原来上移了一行。是根据，一，shape.At(dim)应该不会小于0.  二， 对齐torch，tensor([])[1: 3 : 2]动作，有输出tensor([])，原来的位置则会check住。 torch([1, 2])[1: 3: -1] 同样会被check住，不改变原来的意思。

Comment (hjchen2) [reply oneflow/core/functional/tensor_index.cpp:69]: 加到这里干啥呢，原本是在解析slice的时候就会check的

Comment (hjchen2) [reply oneflow/core/functional/tensor_index.cpp:69]: 这个check还是挪到原来的位置吧，以后说不定要支持负数维度

Comment (hjchen2) [reply oneflow/core/functional/tensor_index.cpp:69]: 好的，这么改没问题

File: oneflow/core/functional/impl/array_functor.cpp [hunk 7]
```diff
@@ -1095,6 +1095,7 @@ class TensorGetItemFunctor {
       for (int i = 0; i < ndims; ++i) {
         if (start[i] != 0 || end[i] != x->shape()->At(i) || step[i] != 1) { return false; }
       }
+      if (target_shape.NumAxes() == 0) { return false; }
```
Comment (wyg1997) [inline oneflow/core/functional/impl/array_functor.cpp:1098]: 这个放在for的上面怎么样？对特殊情况先做特判

Comment (liufengwei0103) [reply oneflow/core/functional/impl/array_functor.cpp:1098]: > 这个放在for的上面怎么样？对特殊情况先做特判

倒是没问题。这里我主要考虑tensor是只有一个元素，又取一个元素的情况。我怕这样改对其他情况产生影响，又想不太明白。

Comment (liufengwei0103) [reply oneflow/core/functional/impl/array_functor.cpp]: > > 这个放在for的上面怎么样？对特殊情况先做特判
> 
> 倒是没问题。这里我主要考虑tensor是只有一个元素，又取一个元素的情况。我怕这样改对其他情况产生影响，又想不太明白。

已修改

File: python/oneflow/test/modules/test_tensor_str.py [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+    test_case.assertTrue("-4.6000" in tensor_str)
+    test_case.assertTrue("60.0000" in tensor_str)
+
+    # scientific representation float dtype
+    tensor = flow.tensor(
+        [[1.3, 2.4, 3.5], [-4.6, 5, 60000000]],
+        device=flow.device(device),
+        dtype=flow.float64,
+    )
+    tensor_str = str(tensor)
```
Comment (jackalcooper) [inline python/oneflow/test/modules/test_tensor_str.py:71]: 这个地方能加一个eval tensor str 的测试吗，类似
```
test_case.assertTrue(eval(tensor_str)==tensor)
```

Comment (wyg1997) [reply python/oneflow/test/modules/test_tensor_str.py:71]: 不太好直接用 == 比较，不过可以比较eval后的device、shape、.numpy() 的结果

Comment (jackalcooper) [reply python/oneflow/test/modules/test_tensor_str.py:71]: 嗯，== 只是个例子，可以用 `array_equal` 或者 `all_close`

Comment (liufengwei0103) [reply python/oneflow/test/modules/test_tensor_str.py]: > 这个地方能加一个eval tensor str 的测试吗，类似
> 
> ```
> test_case.assertTrue(eval(tensor_str)==tensor)
> ```

已增加。

**Review by jackalcooper — APPROVED** [review]

File: python/oneflow/framework/tensor_str.py [hunk 9]
```diff
@@ -34,6 +224,63 @@ def _add_suffixes(tensor_str, suffixes, indent):
...
+        start = [self[i] for i in range(0, PRINT_OPTS.edgeitems)]
+        end = [
+            self[i] for i in range(self.shape[0] - PRINT_OPTS.edgeitems, self.shape[0])
+        ]
+        return flow.stack([get_summarized_data(x) for x in (start + end)])
+    else:
+        return flow.stack([get_summarized_data(x) for x in self])
+
+
+def _gen_tensor_str(inp):
```
Comment (strint) [inline python/oneflow/framework/tensor_str.py:253]: 这里能否出一个没有 data的版本，传入tensor.shape供_gen_tensor_meta_str使用。

另外consistent tensor打印的结果看能不能也给个示例。

Comment (wyg1997) [reply python/oneflow/framework/tensor_str.py:253]: > 这里能否出一个没有 data的版本，传入tensor.shape供_gen_tensor_meta_str使用。

这个有什么特殊的作用吗？主要是要求 eval(str(tensor)) 是可以工作的，如果没有data就不行了

Comment (wyg1997) [reply python/oneflow/framework/tensor_str.py:253]: ### TODO

tensor._meta_repr() -> str
tensor(..., size=[3, 9], 和原有的一致+is_lazy)

File: python/oneflow/framework/tensor_str.py [hunk 10]
```diff
@@ -34,38 +224,74 @@ def _add_suffixes(tensor_str, suffixes, indent):
...
+    if inp.is_consistent:
+        suffixes.append(f"placement={str(inp.placement)}")
+        suffixes.append(f"sbp={str(inp.sbp)}")
+    elif inp.device.type == "cuda":
+        suffixes.append("device='" + str(inp.device) + "'")
+    elif inp.device.type != "cpu":
+        raise RunTimeError("unknow device type")
+    if inp.is_lazy:
         suffixes.append("is_lazy ='True'")
+
```
Comment (wyg1997) [inline python/oneflow/framework/tensor_str.py:269]: 这里 is_lazy 后面多了个空格

Comment (liufengwei0103) [reply python/oneflow/framework/tensor_str.py]: > 这里 is_lazy 后面多了个空格

已修改

**Comment by chengtbf** [comment]
mark。暂时移除 automerge，两个小时以后加回来。

File: python/oneflow/framework/docstr/activation.py [hunk 11]
```diff
@@ -68,7 +68,7 @@
 
         >>> out = flow.F.gelu(input)
         >>> out
-        tensor([-0.1543,  0.    ,  0.3457], dtype=oneflow.float32)
+        tensor([-0.1543,  0.0000,  0.3457], dtype=oneflow.float32)
```
Comment (chengtbf) [inline python/oneflow/framework/docstr/activation.py:71]:
```shell
>>> torch.zeros(2, 2)
tensor([[0., 0.],
        [0., 0.]])
```

torch 并不会 把 0 align 到 0.000，这样做有什么好处吗？ numpy 也不会。

Comment (wyg1997) [reply python/oneflow/framework/docstr/activation.py:71]: torch.zeros(2, 2) 的话我们也是打印 `0.` 的，因为数据都是没有小数位的，这时候会启动整数模式，不打印小数位。

上面这个例子是其它数不满足这个条件，为了格式上对齐，会打印 `0.0000`

Comment (chengtbf) [reply python/oneflow/framework/docstr/activation.py:71]: 确实是这样。学到了~

```shell
>>> x = torch.zeros(3,)
>>> x
tensor([0., 0., 0.])
>>> x[0] = 0.1
>>> x
tensor([0.1000, 0.0000, 0.0000]
```

**Comment by wyg1997** [comment]
好像新合进去一批doctest又出问题了，需要重新测一下Module下所有doctest
